src/sources/timer.rs

```diff
@@ -0,0 +1,538 @@
+//! you can expect precision of order of 1 millisecond, if you need sub-millisecond precision,
+//! make sure you initialize the [`EventLoop`](crate::EventLoop) using
+//! [`EventLoop::try_new_high_precision()`](crate::EventLoop::try_new_high_precision). Note also
+//! that if you need to rely on good precision timers in general, you may need to enable realtime
+//! features of your OS to ensure your thread is quickly woken up by the system scheduler.
+//!
+//! The provided event is an [`Instant`] representing the deadline for which this timer has fired
+//! (which can be earlier than the current time depending on the event loop congestion).
+//!
+//! The callback associated with this event source is expected to return a [`TimeoutAction`], which
+//! can be used to implement self-repeating timers by telling calloop to reprogram the same timer
+//! for a later timeout after it has fired.
+
+use std::{
+    cell::RefCell,
+    collections::BinaryHeap,
+    rc::Rc,
+    sync::{Arc, Mutex},
+    task::Waker,
+    time::{Duration, Instant},
+};
+
+use crate::{EventSource, LoopHandle, Poll, PostAction, Readiness, Token, TokenFactory};
+
+/// A timer event source
+#[derive(Debug)]
+pub struct Timer {
+    registration: Option<(Token, Arc<Mutex<TimerWheel<Token>>>)>,
+    counter: u32,
+    deadline: Instant,
+}
+
+impl Timer {
+    /// Create a timer that will fire immediately when inserted in the event loop
+    pub fn immediate() -> Timer {
+        Self::from_deadline(Instant::now())
+    }
+
+    /// Create a timer that will fire after a given duration from now
+    pub fn from_duration(duration: Duration) -> Timer {
+        Self::from_deadline(Instant::now() + duration)
+    }
+
+    /// Create a timer that will fire at a given instant
+    pub fn from_deadline(deadline: Instant) -> Timer {
+        Timer {
+            registration: None,
+            counter: 0xFFFFFFFF,
+            deadline,
+        }
+    }
+}
+
+impl EventSource for Timer {
+    type Event = Instant;
+    type Metadata = ();
+    type Ret = TimeoutAction;
+    type Error = std::io::Error;
+
+    fn process_events<F>(
+        &mut self,
+        _: Readiness,
+        token: Token,
+        mut callback: F,
+    ) -> Result<PostAction, Self::Error>
+    where
+        F: FnMut(Self::Event, &mut Self::Metadata) -> Self::Ret,
+    {
+        if let Some((my_token, ref wheel)) = self.registration {
+            if my_token != token {
+                return Ok(PostAction::Continue);
+            }
+            let new_deadline = match callback(self.deadline, &mut ()) {
+                TimeoutAction::Drop => return Ok(PostAction::Remove),
+                TimeoutAction::ToInstant(instant) => instant,
+                TimeoutAction::ToDuration(duration) => Instant::now() + duration,
+            };
+            // If we received an event, we MUST have a valid counter value
+            wheel
+                .lock()
```

Comment:
This seems like it could theoretically block the loop. Is there a reason to think it won't?

If so, why not use `try_lock()` and panic if it fails, to effectively `assert!()` that logic?